README.md
# Shorty

Shorty is LearnBoost's URL shortening/redirection service.

## Features

- Redis backed
- Super fast
- At production use on https://lrn.cc
- Uses express, jade, stylus. Easy to hack on!
- Realtime stats with [socket.io](http://socket.io)
    ![](http://f.cl.ly/items/2h2k1p1b2E1I2y0N0Y3u/Image%202011.09.21%208:49:42%20PM.png) 
    ![](http://f.cl.ly/items/072u3V453Q2X0p44180J/Image%202011.09.21%208:16:26%20PM.png)

## Client API

Post to `/` with the `url` field to create.

- If a field is missing or incorrect, status `400` is returned with a JSON
body (`error` key)
- If a problem saving the URL occurs, a status `500` is returned.
- If the url is created, status `200` is returned with JSON body (`short` key)
- If the url already exists, same response as creation is returned.

## Config

To set the short URL to include in JSON responses set the env var
`SHORTY_DOMAIN` to your hostname (e.g.: `lrn.cc`)

## Examples

Using `curl`:

``` shell
curl -X POST "url=http://www.learnboost.com" https://lrn.cc/
```

## Credits

(The MIT License)

Copyright (c) 2011 Guillermo Rauch &lt;guillermo@learnboost.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

### 3rd-party

- Base6k library by Tantek Çelik
- Icon by David Renelt for non-commercial use
(http://www.iconarchive.com/show/little-icon-people-icons-by-david-renelt.html)


History.md

0.0.1 / 2011-10-09
==================

  * Initial release.


public/js/animation.js

/**
 * Triggers a CSS3 keyframe-based animation
 *
 * @param {String} animation name
 * @api public
 */

jQuery.fn.animation = function (name, opts, fn) {
  if ('function' == typeof opts) {
    fn = opts;
    opts = {};
  }

  opts = opts || {};
  opts.duration = opts.duration || '1s';
  opts.timing = opts.timing || 'linear';
  opts.reset = undefined === opts.reset ? true : opts.reset;

  $(this)
    .css('webkitAnimationDuration', opts.duration)
    .css('webkitAnimationTiming', opts.timing)
    .css('webkitAnimationName', name)
    .one('webkitAnimationEnd', function () {
      if (opts.reset) {
        $(this)
          .css('webkitAnimationName', '')
          .css('webkitAnimationDuration', '')
          .css('webkitAnimationTiming', '')
      }
      fn && fn();
    });

  return $(this);
};


app.js

/**
 * Module dependencies.
 */

var express = require('express')
  , stylus = require('stylus')
  , sio = require('socket.io')
  , base60 = require('./base60')
  , jadevu = require('jadevu')
  , url = require('url')
  , nib = require('nib')
  , fs = require('fs')

/**
 * Determine environment.
 */

var env = process.env.NODE_ENV || 'development';

/**
 * Create db.
 */

redis = require('redis').createClient();

/**
 * App creator.
 */

module.exports = (function (port, secure) {

  /**
   * Create app.
   */

  var app;

  if (secure) {
    app = express.createServer({
        key: read('ssl/' + env + '/key.key')
      , cert: read('ssl/' + env + '/cert.crt')
    });
  } else {
    app = express.createServer();
  }

  /**
   * Basic middleware.
   */

  if ('development' == env)  app.use(express.logger('dev'));
  app.use(express.bodyParser());
  app.use(stylus.middleware({ src: __dirname + '/public/', compile: css }));
  app.use(express.static(__dirname + '/public'));

  /**
   * Socket.IO
   */

  var io = sio.listen(app);

  // quiet :)

  io.set('log level', 0);

  /**
   * Reads a file
   *
   * @api private
   */

  function read (file) {
    return fs.readFileSync(__dirname + '/' + file, 'utf8');
  }

  /**
   * Stylus compiler
   */

  function css (str, path) {
    return stylus(str)
      .set('filename', path)
      .set('compress', 'production' == env)
      .use(nib())
      .import('nib');
  };

  /**
   * Configure app.
   */

  app.configure(function () {
    app.set('views', __dirname);
    app.set('view engine', 'jade');
    app.set('domain', process.env.SHORTY_DOMAIN || 'lrn.cc');
  });

  /**
   * Development configuration.
   */

  app.configure('development', function () {
    app.use(express.errorHandler({ dumpExceptions: true, showStack: true }));
  });

  /**
   * Production configuration.
   */

  app.configure('production', function () {
    app.use(express.errorHandler());
  });

  /**
   * GET index page.
   */

  app.get('/', function (req, res, next) {
    redis.hlen('urls', function (err, length) {
      if (err) return next(err);
      res.render('index', { count: length });
    });
  });

  /**
   * POST a url.
   */

  app.post('/', validate, exists, function (req, res, next) {
    var url = req.body.url
      , parsed = req.body.parsed

    // get count of urls
    redis.hlen('urls', function (err, length) {
      if (err) return next(500);

      var short = base60.toString(length ? length + 1 : 0);

      redis.hset('urls', short, url, function (err) {
        if (err) return next(err);
        redis.hset('urls-hash', url, short, function (err) {
          if (err) return next(err);

          var obj = {
              type: 'url created'
            , url: url
            , short: short
            , date: new Date
          };

          redis.lpush('transactions', JSON.stringify(obj), function (err) {
            if (err) return next(500);

            obj.parsed = parsed;
            io.of('/main').volatile.emit('total', length + 1);
            io.of('/stats').volatile.emit('url created', short, parsed, Date.now());
            res.send({ short: 'https://' + app.set('domain') + '/' + short });
          });
        });
      });
    });
  });

  /**
   * Checkes that the URL is valid
   */

  function validate (req, res, next) {
    var parsed = req.body.parsed = url.parse(req.body.url);

    if (!req.body.url || !parsed.protocol || !parsed.host) {
      return res.send(400, { error: 'Bad `url` field' });
    }

    next();
  };

  /**
   * Content negotiation.
   */

  function accept(type) {
    return function(req, res, next){
      if (req.accepts(type)) return next();
      next('route');
    }
  }

  /**
   * Checks that the URL doesnt exist already
   */

  function exists (req, res, next) {
    redis.hget('urls-hash', req.body.url, function (err, val) {
      if (err) return next(err);
      if (val) return res.send({ short: 'https://' + app.set('domain') + '/' + val });
      next();
    });
  }

  /**
   * GET JSON statistics.
   */

  app.get('/stats', accept('json'), function (req, res, next) {
    redis.lrange('transactions', 0, 100, function (err, vals) {
      if (err) return next(err);
      res.send(vals.map(JSON.parse));
    });
  });

  /**
   * GET statistics.
   */

  app.get('/stats', function (req, res, next) {
    redis.lrange('transactions', 0, 100, function (err, vals) {
      if (err) return next(err);
      res.render('stats', { transactions: vals ? vals.map(function (v) {
        v = JSON.parse(v);
        v.parsed = url.parse(v.url);
        delete v.url;
        return v;
      }).reverse() : [] });
    });
  });

  /**
   * GET :short url to perform redirect.
   */

  app.get('/:short', function (req, res, next) {
    redis.hget('urls', req.params.short, function (err, val) {
      if (err) return next(err);
      if (!val) return res.render('404');

      redis.lpush('transactions', JSON.stringify({
          type: 'url visited'
        , url: val
        , short: req.params.short
        , date: Date.now()
        , ip: req.socket.remoteAddress
        , headers: req.headers
      }), function (err) {
        if (err) console.error(err);
      });

      io.of('/stats').volatile.emit(
          'url visited'
        , req.params.short
        , url.parse(val)
        , Date.now()
      );

      res.redirect(val);
    });
  });

  /**
   * Listen.
   */

  if (!module.parent) {
    app.listen(port, function () {
      var addr = app.address();
      console.error(
          '   app listening on ' + addr.address + ':' + addr.port
        + (secure ? ' (secure) ' : '')
      );
    });

    if (!process.listeners('uncaughtException')) {
      process.on('uncaughtException', function (e) {
        console.error(e && e.stack ? e.stack : e);
      });
    }
  }

  return arguments.callee;
})
('production' == env ? 80 : 3000)
('production' == env ? 443 : 3001, true)
